### PR TITLE
[WIP]Custom conflicts for DataSync

### PR DESCRIPTION
**Pull request by ssd71**
This is a very (**very**) rough implementation of custom conflicts in DataSync, need some serious feedback on this before we go ahead with this :smile:
### What is this?
This PR adds the ability for checking custom fields for conflicts and resolving conflicts as well. Similar to what's in Offix conflict-server.

### What's done?
A ConflictEngine interface which could be used to create custom Conflict Engines, not cleanly incorporated into the workflow yet, I am thinking of something like passing a custom conflict engine class to `createDataSyncMongoDbProvider` but I doubt it's optimal because I cannot see any way to extend this other than basic stuff like the engine used by default

### What's left?
- [ ] Finalize approach while optimising for extensiblity
- [ ] Writing tests
- [ ] Writing documentation

### CODE REVIEW

File: packages/graphback-datasync/src/conflict/conflictEngine.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+   */
+  resolveConflicts?(conflict: ConflictStateMap): any
+
+  /**
+   * updateState
+   * Supplies the updated At
+   * @param document document to be updated
+   * @returns document with updated state
+   */
+  updateState?(): FieldTransform[]
```
Comment (wtrocki) [inline packages/graphback-datasync/src/conflict/conflictEngine.ts:30]: This implementation was based on the implementations managing state (for example updating counter), but that is not reliable especially for providers like Mongo. We will not be able to `updateState` here as we actually do that outside using transformer for lastUpdated. 

If we going to do this we might actually need datasync to use it's own updatedAt implementation in updateState. This probably requires discussion but flagging it early

File: packages/graphback-datasync/src/providers/DatasyncMongoDBDataProvider.ts [hunk 2]
```diff
@@ -21,6 +23,17 @@ export class DataSyncMongoDBDataProvider<Type = any> extends MongoDBDataProvider
     ], this.db.collection(this.collectionName)).catch((e: any) => {
       throw e;
     });
+    if (customConflictEngine !== undefined) {
```
Comment (wtrocki) [inline packages/graphback-datasync/src/providers/DatasyncMongoDBDataProvider.ts:26]: Redundant check :D

File: packages/graphback-datasync/src/providers/DatasyncMongoDBDataProvider.ts [hunk 3]
```diff
@@ -21,6 +23,17 @@ export class DataSyncMongoDBDataProvider<Type = any> extends MongoDBDataProvider
...
+      this.conflictEngine = customConflictEngine;
+    } else {
+      this.conflictEngine = new TimestampConflictEngine();
```
Comment (wtrocki) [inline packages/graphback-datasync/src/providers/DatasyncMongoDBDataProvider.ts:29]: Discussion - conflicts enabled by default.
I think we should simply assign value and do not setup default conflict engine. Conflict engine will affect queries etc. so not anyone will want to have it when not needing this feature.

**Comment by wtrocki** [comment]
Looks good so far. I would love to get your opinion about next state - we also need ability to fetch base (I send link to datasync repository that had that implemented) as current production conflict implementation is only client side

File: packages/graphback-datasync/src/providers/DatasyncMongoDBDataProvider.ts (hunk 3 above)
Comment (ssd71) [reply packages/graphback-datasync/src/providers/DatasyncMongoDBDataProvider.ts:29]: Agreed :+1:

File: packages/graphback-datasync/src/conflict/conflictEngine.ts (hunk 1 above)
Comment (ssd71) [reply packages/graphback-datasync/src/conflict/conflictEngine.ts:30]: Yes, I agree, my thoughts on this are pretty jumbled at the moment,
> We will not be able to updateState here as we actually do that outside using transformer for lastUpdated.

I was thinking about just that, what about using a function to transform document here:
https://github.com/aerogear/graphback/pull/1591/files/390b9b642048fddf3b586cc6ce24fd1480f47ce6#diff-be2392f5aaf3d58e02d7e9c6429f279cR151
I think it looks fair, will try it out now, but I have some concerns towards:
> If we going to do this we might actually need datasync to use it's own updatedAt implementation in updateState. 

So, it was kind of a mistake coupling that with datasync as it is vanilla provider stuff I guess, but anyhow datasync still needs updatedAt for delta queries, so what I'm thinking is that for the default strategy with timestamps, we might use a dummy updateState function which will not actually set anything, this way custom conflict engines can update state while not affecting `updatedAt`

What do you think about settling for this until we sort out #1590 ?

**Comment by ssd71** [comment]
> we also need ability to fetch base (I send link to datasync repository that had that implemented) as current production conflict implementation is only client side

Definitely! I am will keep that in mind as well

**Comment by wtrocki** [comment]
I think we need:
- yourself spending another day thinking about this
- have call to decide if conflicts and deltasync are separate or together.

If I add now deltasync my create queries will require timestamp but that is only required for conflicts, which is very confusing. We need to clear this as soon as possible.

Technical capabilities will follow - at first this needs to make sense from the functionality point of view

File: packages/graphback-datasync/src/DataSyncPlugin.ts [hunk 4]
```diff
@@ -85,6 +87,10 @@ export class DataSyncPlugin extends GraphbackPlugin {
       }
     })
 
+    if (dataSyncModelCount === 0) {
+      console.error("No DataSync Models detected, ensure that your models are properly annotated.")
```
Comment (craicoverflow) [inline packages/graphback-datasync/src/DataSyncPlugin.ts:91]: Maybe use `this.logWarning` to keep consistent with the other plugins ([including this one](https://github.com/aerogear/graphback/blob/071c2744aed20eea9414ac9c8e445c620e1545d9/packages/graphback-datasync/src/DataSyncPlugin.ts#L31))

**Comment by ssd71** [comment]
Branch is stale now, going to push prototype for this in the coming days

**Pull request closed without merging**